### PR TITLE
Add 'ignore_index' input in the spec for SoftmaxCrossEntropyLoss and NLLLoss.

### PR DESCRIPTION
**Pull request by codemzs**
This change fixes the spec for SoftmaxCrossEntropyLoss and NLLLoss to make it compatible with PyTorch.

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -14814,6 +14814,8 @@ This version of the operator has been available since version 12 of the default
 <dd>Target tensor of shape (N) or (N, d1, d2, ..., dk). Target element value shall be in range of [0, C).</dd>
 <dt><tt>weight</tt> (optional) : T</dt>
 <dd>Optional rescaling weight tensor. If given, it has to be a tensor of size C. Otherwise, it is treated as if having all ones.</dd>
+<dt><tt>ignore_index</tt> (optional) : Tind</dt>
```
Comment (wschin) [inline docs/Changelog.md]: If we can set the `weight` of a single class to 0, why do we need `ignore_index`?

Comment (codemzs) [reply docs/Changelog.md]: Hacking around is not good. You will increase the computation cost with this workaround. We are keeping this spec strictly for PyTorch. Next version will have a provision for TensorFlow.

**Comment by lara-hdr** [comment]
@codemzs, thanks for the PR :)
I have some comments but I cannot add them inline since the lines are not modified;

NllLoss and SoftmaxCrossEntropy are functions not operators; so they have a function body in ONNX that we will need to update, even if there is a kernel.

The functions are defined in onnx/defs/math/defs.cc, line 1831 and 2320;
We will need to update the 'if' clauses 'if (!ctx.hasInput(2))' to deal with cases with 2, 3, 4 inputs.In case ignore_iindex is specified (4 inputs), we can use ScatterElements to put a 0 in the weight on the ignored index in the weights.

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -2367,38 +2373,44 @@ ONNX_OPERATOR_SET_SCHEMA(
...
             "K-dimensional loss. Otherwise, it is a scalar.",
             "T")
-	.Output(
-	    1,
-	    "log_prob",
-	    "Log probability tensor. If the output of softmax is prob, its value is log(prob).",
-	    "T",
-	    OpSchema::Optional)
-        .TypeConstraint(
+        .Output(
```
Comment (lara-hdr) [inline onnx/defs/math/defs.cc]: why is this is required now?
PyTorch only has 1 output for NllLoss/CrossEntropy

File: onnx/backend/test/case/node/softmaxcrossentropy.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
 from . import expect
 
 
-def softmaxcrossentropy(x, target, weight=None, reduction='mean'):  # type: ignore
+def softmaxcrossentropy(x, target, weight=None, reduction='mean', ignore_index=-100):  # type: ignore
```
Comment (lara-hdr) [inline onnx/backend/test/case/node/softmaxcrossentropy.py]: maybe set ignore_index to None?
(Unlike PyTorch, the spec does not say that the default is -100.)

File: onnx/backend/test/case/node/softmaxcrossentropy.py [hunk 4]
```diff
@@ -178,3 +180,27 @@ def export_softmaxcrossentropy_mean_weights():  # type: () -> None
 
         # Check results
         expect(node, inputs=[x, labels, weights], outputs=[sce], name='test_softmax_cross_entropy_mean_weight')
+
+    @staticmethod
+    def export_softmaxcrossentropy_mean_weights_ignore_index():  # type: () -> None
```
Comment (lara-hdr) [inline onnx/backend/test/case/node/softmaxcrossentropy.py]: can we add a case with no weights and ignore_index specified?

File: onnx/backend/test/case/node/softmaxcrossentropy.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 from . import expect
```
Comment (lara-hdr) [inline onnx/backend/test/case/node/softmaxcrossentropy.py]: Should the nllloss tests be updated?

**Comment by snnn** [comment]
Is there any tests to verify if the function body works as expected?

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (KsenijaS) [reply onnx/defs/math/defs.cc]: it's optional output for softmaxcrossentropy, it was a part of the original PR

File: docs/Changelog.md (hunk 1 above)
Comment (gramalingam) [reply docs/Changelog.md]: I don't understand this. Using weights seems the better "non-hacking" solution to me. We already have weights. Is it aligned to PyTorch or TF or does it handle both? If the weights doesn't correctly handle both PyTorch and TF, we already have a problem that needs to be fixed.

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (lara-hdr) [reply onnx/defs/math/defs.cc]: @KsenijaS initially it was optional, but this PR makes it required.
I am curious of the reason since PyTorch does not have it as an output

File: onnx/defs/math/defs.cc [hunk 6]
```diff
@@ -1897,6 +1897,12 @@ ONNX_OPERATOR_SET_SCHEMA(
             "If given, it has to be a tensor of size C. Otherwise, it is treated as if having all ones.",
             "T",
             OpSchema::Optional)
+        .Input(
```
Comment (lara-hdr) [inline onnx/defs/math/defs.cc]: why not make it an attribute?

Comment (lara-hdr) [reply onnx/defs/math/defs.cc]: Talked with @codemzs offline, he is ok to make this an attribute for NllLoss and SoftmaxCrossEntropy.
cc : @SherlockNoMad

**Comment by gramalingam** [comment]
> Is there any tests to verify if the function body works as expected?

Every "node" test-case is also expanded into a test-case for the corresponding function-body: for example, like here: https://github.com/onnx/onnx/tree/master/onnx/backend/test/data/node/test_softmax_cross_entropy_mean_3d_expanded

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (lara-hdr) [reply onnx/defs/math/defs.cc]: my bad, it is still optional; the "OpSchema::Optional" is after the below red lines.

File: docs/Changelog.md [hunk 7]
```diff
@@ -14801,6 +14801,8 @@ This version of the operator has been available since version 12 of the default
 #### Attributes
 
 <dl>
+<dt><tt>ignore_index</tt> : int (default is -100)</dt>
```
Comment (wschin) [inline docs/Changelog.md]: Where does this default come from? If we don't have a good default value, maybe we can make this attribute non-optional?

File: docs/Operators.md [hunk 8]
```diff
@@ -18394,6 +18398,37 @@ expect(node, inputs=[x, labels, weights], outputs=[sce], name='test_softmax_cros
...
+node = onnx.helper.make_node('SoftmaxCrossEntropyLoss',
+                             inputs=['x', 'y', 'w'],
+                             outputs=['z'],
+                             reduction=reduction,
+                             ignore_index=ignore_index)
+
+# Define operator inputs.
+np.random.seed(0)
+x = np.random.rand(3, 5).astype(np.float32)
+labels = np.random.randint(0, high=5, size=(3, ))
```
Comment (wschin) [inline docs/Operators.md]: Maybe you want to make sure this `labels` contain `0`, which will be ignored.

File: onnx/backend/test/case/node/softmaxcrossentropy.py [hunk 9]
```diff
@@ -20,13 +20,15 @@ def softmaxcrossentropy(x, target, weight=None, reduction='mean'):  # type: igno
         N, C = input_shape
         neg_gather_element_input = np.zeros((N, ), dtype=np.float32)
         for i in range(N):
-            neg_gather_element_input[i] = -inp[i][target[i]]
+            if target[i] != ignore_index:
```
Comment (wschin) [inline onnx/backend/test/case/node/softmaxcrossentropy.py]: Just want to double-check. Do you have some python code to compare pytorch and this reference implementation?

File: docs/Changelog.md (hunk 1 above)
Comment (codemzs) [reply docs/Changelog.md]: Resolved offline with @wschin. In his words "I understand composing Pytorch SoftmaxCrossEntropy is hard (but still doable with ops including LabelEncoder). I am fine with this change because using LabelEncoder doesn’t sound a good idea. I feel we can still keep in mind that ONNX is an IR for both of Tensorflow and Pytorch. If we only consider Pytorch, the value of ONNX may decrease."

Yes, this spec also supports TensorFlow’s tf.nn.sparse_softmax_cross_entropy_with_logits but we will need a minor change in the TensorFlow converter to permute and reshape the logits from [N, D1, D2…Dk, C] to [N, C, D1, D2…Dk]

File: docs/Changelog.md (hunk 7 above)
Comment (codemzs) [reply docs/Changelog.md]: It is coming from PyTorch but I can make it non-optional.

**Comment by codemzs** [comment]
@wschin I have removed the default value for `ignore_index` but still kept it as optional because it is easier to implement the function body. In the converter we will pass an `ignore_index` only if its value provided is within the range [0, C). I did this at the suggestion of @lara-hdr.

File: onnx/backend/test/case/node/softmaxcrossentropy.py (hunk 4 above)
Comment (codemzs) [reply onnx/backend/test/case/node/softmaxcrossentropy.py]: We spoke offline, now that ignore_index is not an input so this is not needed.

File: onnx/backend/test/case/node/softmaxcrossentropy.py (hunk 9 above)
Comment (lara-hdr) [reply onnx/backend/test/case/node/softmaxcrossentropy.py]: Yes, here it is.
[sce_test.zip](https://github.com/onnx/onnx/files/4400389/sce_test.zip)

Comment (codemzs) [reply onnx/backend/test/case/node/softmaxcrossentropy.py]: thanks @lara-hdr for running these tests and validating them with PyTorch.

File: onnx/defs/math/defs.cc [hunk 10]
```diff
@@ -1939,7 +1939,8 @@ ONNX_OPERATOR_SET_SCHEMA(
             std::string("mean"))
         .Attr(
             "ignore_index",
-            "Specifies a target value that is ignored and does not contribute to the input gradient.",
+            "Specifies a target value that is ignored and does not contribute to the input gradient. "
+            "It is an optional value and valid values are [0, C).",
```
Comment (snnn) [inline onnx/defs/math/defs.cc]: Usually the optional attribute should have a default value defined in the schema, unless the default value is too complex to be provided there.

Comment (codemzs) [reply onnx/defs/math/defs.cc]: Please read @wschin 's comment. ORT also has a way to get default value when one is not defined, i.e `GetAttrOrDefault`

Comment (snnn) [reply onnx/defs/math/defs.cc]: Thank you for clarifying this.

File: onnx/defs/math/defs.cc [hunk 11]
```diff
@@ -1837,33 +1843,66 @@ bool BuildContextDependentFunctionBody(const FunctionBodyBuildContext& ctx, cons
...
-        body.push_back({{"loss"}, "ReduceSum", {"loss_Ndd"}, {MakeAttribute("keepdims", (int64_t)0)}});  
-      }
-    } 
-  } else {
-    body.push_back({{"weight_gather"}, "Gather", {"weight", "target"}});
-    body.push_back({{"loss_unweighted"}, "Squeeze", {"loss_N1dd"}, {MakeAttribute("axes", std::vector<int64_t>({1}))}});
-    if (ctx.getAttribute("reduction")->s() == "none") {
-      body.push_back({{"loss"}, "Mul", {"loss_unweighted", "weight_gather"}});
+    if(ctx.getAttribute("ignore_index") == nullptr)
+    {
```
Comment (snnn) [inline onnx/defs/math/defs.cc]: Nit: please have a consistent coding style. (The position of "{" )

**Comment by codemzs** [comment]
@wschin / @ebarsoum / @gramalingam Please merge this change. All of the **SoftmaxCrossEntropyLoss** & **NegativeLogLikelihoodLoss** ONNX tests on ONNX Runtime are now passing with the updated spec. @lara-hdr  has also verified the outputs match with that of PyTorch. 

image

Thank you, @chinhuang007 for giving me time till end of weekend to wrap this change up, much appreciated! This change is ready to be shipped with ONNX 1.7 release.

CC: @SherlockNoMad

**Comment by chinhuang007** [comment]
@codemzs Thank you for the contribution!
